src/clusterfuzz/_internal/bot/tasks/utasks/uworker_io.py

```diff
@@ -0,0 +1,245 @@
+  GCS url for the tworker to upload it (this happens first)."""
+  gcs_path = get_uworker_io_gcs_path()
+  return storage.get_signed_download_url(gcs_path), gcs_path
+
+
+def upload_uworker_input(uworker_input):
+  """Uploads input for the untrusted portion of a task."""
+  signed_download_url, gcs_path = get_uworker_input_urls()
+
+  with tempfile.NamedTemporaryFile() as uworker_input_file:
+    with open(uworker_input_file.name, 'w') as fp:
+      fp.write(uworker_input)
+    if not storage.copy_file_to(uworker_input_file.name, gcs_path):
+      raise RuntimeError('Failed to upload uworker_input.')
+  return signed_download_url
+
+
+def make_ndb_entity_input_obj_serializable(obj):
+  """Returns a dictionary that can be JSON serialized representing an NDB
+  entity. Does not include datetime fields."""
+  obj_dict = obj.to_dict()
+  # TOOD(metzman): Handle datetimes.
+  for key in list(obj_dict.keys()):
+    value = obj_dict[key]
+    if isinstance(value, datetime.datetime):
+      del obj_dict[key]
+  return {
+      'key': base64.b64encode(obj.key.serialized()).decode(),
+      'properties': obj_dict,
+  }
+
+
+def get_entity_with_properties(ndb_key: ndb.Key, properties) -> ndb.Model:
+  """Returns the entity pointed to by ndb_key and sets the properties on the
+  entity as the |properties| dictionary specifies."""
+  model_name = ndb_key.kind()
+  model_cls = getattr(data_types, model_name)
+  entity = model_cls()
+  entity.key = ndb_key
+  for ndb_property, value in properties.items():
+    fail_msg = f'{entity} doesn\'t have {ndb_property}'
+    assert hasattr(entity, ndb_property), fail_msg
+    setattr(entity, ndb_property, value)
+  return entity
+
+
+def deserialize_uworker_input(serialized_uworker_input):
+  """Deserializes input for the untrusted part of a task."""
+  serialized_uworker_input = json.loads(serialized_uworker_input)
+  uworker_input = serialized_uworker_input['serializable']
+  for name, entity_dict in serialized_uworker_input['entities'].items():
+    entity_key = entity_dict['key']
+    serialized_key = base64.b64decode(bytes(entity_key, 'utf-8'))
+    ndb_key = ndb.Key(serialized=serialized_key)
+    entity = get_entity_with_properties(ndb_key, entity_dict['properties'])
+    uworker_input[name] = UworkerEntityWrapper(entity)
+  return uworker_input
+
+
+def serialize_uworker_input(uworker_input):
+  """Serializes and returns |uworker_input| as JSON. Can handle ndb entities."""
+  serializable = {}
+  ndb_entities = {}
+  for key, value in uworker_input.items():
+    if not isinstance(value, ndb.Model):
+      serializable[key] = value
+      continue
+    ndb_entities[key] = make_ndb_entity_input_obj_serializable(value)
+
+  return json.dumps({'serializable': serializable, 'entities': ndb_entities})
+
+
+def serialize_and_upload_uworker_input(uworker_input, job_type,
+                                       uworker_output_upload_url) -> str:
```

Comment:
I am in favor of type annotations like `-> str` at the end of this function.
Would it be OK if I use it in my code too?

Reply:
It's OK if you do it.